In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn import datasets
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn import linear_model,metrics
from sklearn.preprocessing  import StandardScaler

# Data discription
- Each row of the dataset is a single game with the following features (in the order in the vector):
- Team won the game (1 or -1)
- Cluster ID (related to location)
- Game mode (eg All Pick)
- Game type (eg. Ranked)
- end: Each element is an indicator for a hero. Value of 1 indicates that a player from team '1' played as that hero and '-1' for the other team. Hero can be selected by only one player each game. This means that each row has five '1' and five '-1' values.

# Problem
- 判斷這個隊伍有沒有獲勝(第一個Attribute為1)

# data preprocessing
- 利用Cluster ID、遊戲模式、遊戲類型去判斷隊伍輸或贏

In [3]:
df_train = pd.read_csv('./dota2Train.csv')
df_test = pd.read_csv('./dota2Test.csv')

In [103]:
train_X = df_train.iloc[:, 1:4].reset_index(drop=True)
train_X.head()

223  2  2.1
0  152  2    2
1  131  2    2
2  154  2    2
3  171  2    3
4  122  2    3

In [104]:
train_Y = df_train.iloc[:,[0]]
train_Y.head()

-1
0   1
1   1
2   1
3  -1
4   1

In [105]:
test_X = df_test.iloc[:, 1:4].reset_index(drop=True)
test_X.head()

223  8  2
0  227  8  2
1  136  2  2
2  227  2  2
3  184  2  3
4  231  2  2

In [106]:
test_Y = df_test.iloc[:,[0]]
test_Y.head()

-1
0   1
1  -1
2   1
3   1
4   1

In [107]:
sc = StandardScaler()
sc.fit(train_X)

x_train=sc.transform(train_X)
x_test=sc.transform(test_X)

C:\Users\Jerry\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\Jerry\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  after removing the cwd from sys.path.
C:\Users\Jerry\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  """


In [108]:
mlp = MLPClassifier(solver='lbfgs', alpha=5e-2,hidden_layer_sizes=(5, 2), random_state=1)
mlp.fit(x_train, train_Y)
mlp.score(x_test, test_Y)

C:\Users\Jerry\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.5345380355581463

# Result
- Training的結果為53.5%

# Data preprocessing2
- 利用所選取的英雄去判斷輸贏

In [109]:
train_X = df_train.iloc[:, 4:].reset_index(drop=True)
train_Y = df_train.iloc[:,[0]]
train_X.head()

0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  1 ...   0.94  0.95  0.96  0.97  \
0  0    0    0    1    0   -1    0    0    0  0 ...      0     0     0     0   
1  0    0    0    1    0   -1    0    0    0  0 ...      0     0     0     0   
2  0    0    0    0    0    0   -1    0    0  0 ...      0     0     0     0   
3  0    0    0    0    0   -1    0    0   -1  0 ...      0     0     0     0   
4  0    1    0    0    0    0    0    0    0  0 ...      0     0     0     0   

   0.98  0.99  0.100  0.101  0.102  top  
0     0     0      0      0      0    0  
1     0     0      0      0      0    0  
2     0     0      0      0      0    0  
3     0     0      0      0      0    0  
4     0     0      0      0     -1    0  

[5 rows x 114 columns]

In [110]:
test_X = df_test.iloc[:, 4:].reset_index(drop=True)
test_Y = df_test.iloc[:,[0]]
test_X.head()

0  -1.1  0.1  0.2  0.3  0.4  0.5  0.6  1  0.7 ...   0.94  0.95  0.96  0.97  \
0  0     0    0    0    0    0    0    0  1    0 ...      0     0     0     0   
1  1     0    0    0   -1    0    0    0  1    1 ...      0     0     0     0   
2 -1     0    0    0    0    0    0    0  1    0 ...      0     0     0     0   
3  0     0    0   -1    0    0    0   -1  0    0 ...      0     0     0     0   
4  0     0    0    0    0    0    0    0 -1    0 ...      0     0     0     0   

   0.98  0.99  0.100  0.101  0.102  top  
0     0     0      0      0      0    1  
1     0     0      0      0      0    1  
2     0     0      0      0      0    1  
3     0     0      0      0      0    0  
4     0    -1      0      0      0    0  

[5 rows x 114 columns]

In [111]:
sc = StandardScaler()
sc.fit(train_X)

x_train=sc.transform(train_X)
x_test=sc.transform(test_X)

C:\Users\Jerry\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\Jerry\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  after removing the cwd from sys.path.
C:\Users\Jerry\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  """


In [112]:
mlp = MLPClassifier(solver='lbfgs', alpha=5e-2,hidden_layer_sizes=(5, 2), random_state=1)
mlp.fit(x_train, train_Y)
mlp.score(x_test, test_Y)

C:\Users\Jerry\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.5927329252890314

# Result
- Training結果為59.3%

# Conclusion 
- 我分成兩類，一類為Cluster ID、對戰模式以及遊戲形態，另一類為英雄對戰組合。
- 我試過將Cluster ID獨立出來後Training，結果不變，因此可看出在第一種data preprocessing下，主要控制的因素在於對戰模式以及遊戲形態，與玩家無關。
- 第二種的準確度比較高，表示對戰組合導致的勝負結果會比較準，但準確性也才59.3%。

# More information
- 從這個資料裡還可以得到更多的資訊，例如在這個賽季中最常被使用的英雄、或是最常出現的對戰組合，進而分析強勢的英雄種類。
- 相反地，根據出場率最低的英雄，可讓遊戲公司去做角色的微調。

In [77]:
n = 0
freqency = []
for i in range(4,117):
    for j in range(0,10000):
        if df_train.iloc[j,i]==1:
            n = n+1
    freqency.append(n)
    n=0
print(freqency)

[802, 1014, 139, 655, 505, 1200, 627, 1235, 1720, 575, 533, 370, 242, 1529, 226, 585, 611, 923, 411, 401, 778, 509, 495, 0, 407, 642, 331, 585, 352, 362, 239, 1089, 242, 615, 944, 595, 204, 138, 356, 251, 882, 670, 233, 1659, 167, 263, 356, 411, 263, 300, 200, 140, 465, 673, 166, 402, 459, 181, 672, 246, 111, 681, 431, 178, 197, 45, 401, 350, 171, 535, 547, 180, 661, 1181, 428, 229, 119, 68, 308, 85, 251, 195, 144, 516, 213, 419, 258, 414, 158, 253, 80, 42, 1232, 433, 324, 246, 306, 906, 419, 216, 239, 227, 293, 1387, 232, 413, 122, 0, 238, 237, 139, 188, 109]


In [80]:
print(np.where(freqency==np.max(freqency)))
print(np.where(freqency==np.min(freqency)))

(array([8], dtype=int64),)
(array([ 23, 107], dtype=int64),)


# Result
- 第9號英雄出場率為最高，因此可能為此賽季的強勢角色。
- 第24、108號英雄出場率最低，因此為弱勢角色，遊戲公司可參考並調整此角色。

# Improvement
- 藉由強勢英雄，可以去分析當隊伍拿到強勢英雄時，勝率應該會提高，加入這項attribute再分析，讓準確度提高。

In [83]:
top_hero=[]
for i in range(0,92649):
    if df_train.iloc[i,12]==1:
        top_hero.append(1)
    else:
        top_hero.append(0)
df_train['top']= top_hero
df_train.head()

-1  223  2  2.1  0  0.1  0.2  0.3  0.4  0.5 ...   0.94  0.95  0.96  0.97  \
0   1  152  2    2  0    0    0    1    0   -1 ...      0     0     0     0   
1   1  131  2    2  0    0    0    1    0   -1 ...      0     0     0     0   
2   1  154  2    2  0    0    0    0    0    0 ...      0     0     0     0   
3  -1  171  2    3  0    0    0    0    0   -1 ...      0     0     0     0   
4   1  122  2    3  0    1    0    0    0    0 ...      0     0     0     0   

   0.98  0.99  0.100  0.101  0.102  top  
0     0     0      0      0      0    0  
1     0     0      0      0      0    0  
2     0     0      0      0      0    0  
3     0     0      0      0      0    0  
4     0     0      0      0     -1    0  

[5 rows x 118 columns]

In [89]:
top_hero=[]
for i in range(0,10293):
    if df_test.iloc[i,12]==1:
        top_hero.append(1)
    else:
        top_hero.append(0)
df_test['top']= top_hero
df_test.head()

-1  223  8  2  0  -1.1  0.1  0.2  0.3  0.4 ...   0.94  0.95  0.96  0.97  \
0   1  227  8  2  0     0    0    0    0    0 ...      0     0     0     0   
1  -1  136  2  2  1     0    0    0   -1    0 ...      0     0     0     0   
2   1  227  2  2 -1     0    0    0    0    0 ...      0     0     0     0   
3   1  184  2  3  0     0    0   -1    0    0 ...      0     0     0     0   
4   1  231  2  2  0     0    0    0    0    0 ...      0     0     0     0   

   0.98  0.99  0.100  0.101  0.102  top  
0     0     0      0      0      0    1  
1     0     0      0      0      0    1  
2     0     0      0      0      0    1  
3     0     0      0      0      0    0  
4     0    -1      0      0      0    0  

[5 rows x 118 columns]

In [94]:
train_X = df_train.iloc[:, 1:].reset_index(drop=True)
train_Y = df_train.iloc[:,[0]]
train_X.head()

223  2  2.1  0  0.1  0.2  0.3  0.4  0.5  0.6 ...   0.94  0.95  0.96  0.97  \
0  152  2    2  0    0    0    1    0   -1    0 ...      0     0     0     0   
1  131  2    2  0    0    0    1    0   -1    0 ...      0     0     0     0   
2  154  2    2  0    0    0    0    0    0   -1 ...      0     0     0     0   
3  171  2    3  0    0    0    0    0   -1    0 ...      0     0     0     0   
4  122  2    3  0    1    0    0    0    0    0 ...      0     0     0     0   

   0.98  0.99  0.100  0.101  0.102  top  
0     0     0      0      0      0    0  
1     0     0      0      0      0    0  
2     0     0      0      0      0    0  
3     0     0      0      0      0    0  
4     0     0      0      0     -1    0  

[5 rows x 117 columns]

In [95]:
test_X = df_test.iloc[:, 1:].reset_index(drop=True)
test_Y = df_test.iloc[:,[0]]
test_X.head()

223  8  2  0  -1.1  0.1  0.2  0.3  0.4  0.5 ...   0.94  0.95  0.96  0.97  \
0  227  8  2  0     0    0    0    0    0    0 ...      0     0     0     0   
1  136  2  2  1     0    0    0   -1    0    0 ...      0     0     0     0   
2  227  2  2 -1     0    0    0    0    0    0 ...      0     0     0     0   
3  184  2  3  0     0    0   -1    0    0    0 ...      0     0     0     0   
4  231  2  2  0     0    0    0    0    0    0 ...      0     0     0     0   

   0.98  0.99  0.100  0.101  0.102  top  
0     0     0      0      0      0    1  
1     0     0      0      0      0    1  
2     0     0      0      0      0    1  
3     0     0      0      0      0    0  
4     0    -1      0      0      0    0  

[5 rows x 117 columns]

In [96]:
sc = StandardScaler()
sc.fit(train_X)

x_train=sc.transform(train_X)
x_test=sc.transform(test_X)

C:\Users\Jerry\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:625: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\Jerry\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  after removing the cwd from sys.path.
C:\Users\Jerry\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype int64 were all converted to float64 by StandardScaler.
  """


In [97]:
mlp = MLPClassifier(solver='lbfgs', alpha=5e-2,hidden_layer_sizes=(5, 2), random_state=1)
mlp.fit(x_train, train_Y)
mlp.score(x_test, test_Y)

C:\Users\Jerry\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:916: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.5976877489556008

# Conclusion
- Training結果為59.8%
- 沒有提高太多的原因，我認為是因為我只挑了一種最常選的英雄(top hero)，但遊戲是5 vs 5的，因此如果挑選top 5應該會再更好一點。
- 不過整體來講，我覺得以遊戲上來說即使對戰組合一樣(選取英雄相同)，也有極大的機率產生不同的結果，因此Training的結果不太準確。
- 如果能夠再加上一些額外的資訊，像是兩個戰隊的積分、或是戰隊的等級等等，應該更有幫助去分析對戰的輸贏結果。